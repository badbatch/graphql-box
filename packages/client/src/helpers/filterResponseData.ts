import {
  FragmentDefinitionNodeMap,
  MaybeResponseData,
  PlainObjectMap,
  QUERY,
  RequestContext,
  RequestData,
  ResponseData,
} from "@graphql-box/core";
import {
  KeysAndPathsOptions,
  buildFieldKeysAndPaths,
  getAliasOrName,
  getChildFields,
  getFragmentDefinitions,
  getOperationDefinitions,
  hasChildFields,
  iterateChildFields,
} from "@graphql-box/helpers";
import { FieldNode } from "graphql";
import { get, isArray, isPlainObject, isUndefined, set } from "lodash";
import { FilteredDataAndCacheMetadata } from "../defs";

export const filterDataAndCacheMetadata = (
  pendingFieldNode: FieldNode,
  activeFieldNode: FieldNode,
  activeResponseData: MaybeResponseData,
  { filteredData, filteredCacheMetadata }: FilteredDataAndCacheMetadata,
  fragmentDefinitions: { active?: FragmentDefinitionNodeMap; pending?: FragmentDefinitionNodeMap },
  keyAndPathOptions: { active: KeysAndPathsOptions; pending: KeysAndPathsOptions },
  contexts: { active: RequestContext; pending: RequestContext },
) => {
  const pendingKeysAndPaths = buildFieldKeysAndPaths(pendingFieldNode, keyAndPathOptions.active, contexts.pending);
  const activeKeysAndPaths = buildFieldKeysAndPaths(activeFieldNode, keyAndPathOptions.active, contexts.active);

  if (activeKeysAndPaths.requestFieldCacheKey !== pendingKeysAndPaths.requestFieldCacheKey) {
    return;
  }

  const rawFieldData = get(activeResponseData.data, activeKeysAndPaths.responseDataPath);
  let activeFieldData = rawFieldData;

  if (isPlainObject(activeFieldData)) {
    activeFieldData = {};
  } else if (isArray(activeFieldData)) {
    activeFieldData = [];
  }

  set(filteredData, pendingKeysAndPaths.responseDataPath, activeFieldData);
  const cacheability = activeResponseData.cacheMetadata?.get(activeKeysAndPaths.requestFieldPath);

  if (cacheability) {
    filteredCacheMetadata.set(pendingKeysAndPaths.requestFieldPath, cacheability);
  }

  if (!hasChildFields(pendingFieldNode, { fragmentDefinitions: fragmentDefinitions.pending })) {
    return;
  }

  iterateChildFields(
    pendingFieldNode,
    rawFieldData,
    fragmentDefinitions.pending,
    (
      pendingChildFieldNode: FieldNode,
      _childTypeName: string | undefined,
      _childFragmentKind: string | undefined,
      _childFragmentName: string | undefined,
      childIndex?: number,
    ) => {
      let activeChildFieldNode: FieldNode;

      if (isUndefined(childIndex)) {
        const matchingActiveFieldAndTypeName = getChildFields(activeFieldNode, {
          fragmentDefinitions: fragmentDefinitions.active,
          name: getAliasOrName(pendingChildFieldNode),
        });

        if (!matchingActiveFieldAndTypeName?.length) {
          return;
        }

        activeChildFieldNode = matchingActiveFieldAndTypeName[0].fieldNode;
      } else {
        activeChildFieldNode = activeFieldNode;
      }

      filterDataAndCacheMetadata(
        pendingChildFieldNode,
        activeChildFieldNode,
        activeResponseData,
        { filteredData, filteredCacheMetadata },
        fragmentDefinitions,
        {
          active: { ...activeKeysAndPaths, index: childIndex },
          pending: { ...pendingKeysAndPaths, index: childIndex },
        },
        contexts,
      );
    },
  );
};

export default (
  pendingRequestDatat: RequestData,
  activeRequestDatat: RequestData,
  { data, cacheMetadata, ...rest }: ResponseData,
  { active, pending }: { active: RequestContext; pending: RequestContext },
) => {
  const pendingQueryNode = getOperationDefinitions(pendingRequestDatat.ast, QUERY)[0];
  const pendingQueryFragmentDefinitions = getFragmentDefinitions(pendingRequestDatat.ast);
  const activeQueryNode = getOperationDefinitions(activeRequestDatat.ast, QUERY)[0];
  const activeQueryFragmentDefinitions = getFragmentDefinitions(activeRequestDatat.ast);

  const pendingFieldsAndTypeNames = getChildFields(pendingQueryNode, {
    fragmentDefinitions: pendingQueryFragmentDefinitions,
  });

  const filteredData: PlainObjectMap = {};
  const filteredCacheMetadata = new Map();

  pendingFieldsAndTypeNames?.forEach(({ fieldNode: pendingFieldNode }) => {
    const matchingActiveFieldAndTypeName = getChildFields(activeQueryNode, {
      fragmentDefinitions: activeQueryFragmentDefinitions,
      name: getAliasOrName(pendingFieldNode),
    });

    if (!matchingActiveFieldAndTypeName) {
      return;
    }

    const { fieldNode: activeFieldNode } = matchingActiveFieldAndTypeName[0];

    filterDataAndCacheMetadata(
      pendingFieldNode,
      activeFieldNode,
      { data, cacheMetadata, ...rest },
      { filteredData, filteredCacheMetadata },
      {
        active: activeQueryFragmentDefinitions,
        pending: pendingQueryFragmentDefinitions,
      },
      {
        active: { requestFieldPath: QUERY },
        pending: { requestFieldPath: QUERY },
      },
      { active, pending },
    );
  });

  return {
    ...rest,
    cacheMetadata: filteredCacheMetadata,
    data: filteredData,
  };
};
