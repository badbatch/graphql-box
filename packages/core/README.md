# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
