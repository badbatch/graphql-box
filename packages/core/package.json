{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}