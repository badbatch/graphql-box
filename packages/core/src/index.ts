import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
