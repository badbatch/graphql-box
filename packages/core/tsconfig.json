{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}