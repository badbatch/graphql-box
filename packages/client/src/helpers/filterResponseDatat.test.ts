import { type ResponseData } from '@graphql-box/core';
import {
  getRequestContext,
  getRequestData,
  parsedRequests,
  requestFieldTypeMaps,
  responses,
} from '@graphql-box/test-utils';
import { expect } from '@jest/globals';
import { filterResponseData } from './filterResponseData.ts';

describe('filterResponseData', () => {
  describe('when there are no fragments or directives or aliases in the pending query', () => {
    it('should return the correct data', () => {
      const pendingRequestData = getRequestData(parsedRequests.singleTypeQuery);
      const pendingRequestContext = getRequestContext({ fieldTypeMap: requestFieldTypeMaps.singleTypeQuery });
      const activeRequestData = getRequestData(parsedRequests.nestedTypeQuery);
      const activeRequestContext = getRequestContext({ fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery });

      const responseData = filterResponseData(
        pendingRequestData,
        activeRequestData,
        responses.nestedTypeQuery as ResponseData,
        {
          active: activeRequestContext,
          pending: pendingRequestContext,
        }
      );

      expect(responseData).toMatchSnapshot();
    });
  });

  describe('when there is an alias in the pending query', () => {
    it('should return the correct data', () => {
      const pendingRequestData = getRequestData(parsedRequests.singleTypeQueryWithAlias);
      const pendingRequestContext = getRequestContext({ fieldTypeMap: requestFieldTypeMaps.singleTypeQuery });
      const activeRequestData = getRequestData(parsedRequests.nestedTypeQuery);
      const activeRequestContext = getRequestContext({ fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery });

      const responseData = filterResponseData(
        pendingRequestData,
        activeRequestData,
        responses.nestedTypeQuery as ResponseData,
        {
          active: activeRequestContext,
          pending: pendingRequestContext,
        }
      );

      expect(responseData).toMatchSnapshot();
    });
  });

  describe('when there is a directive in the pending query', () => {
    describe('when the directives do not match', () => {
      it('should return the correct data', () => {
        const pendingRequestData = getRequestData(parsedRequests.singleTypeQueryWithDirective);

        const pendingRequestContext = getRequestContext({
          fieldTypeMap: requestFieldTypeMaps.singleTypeQueryWithDirective,
        });

        const activeRequestData = getRequestData(parsedRequests.nestedTypeQuery);
        const activeRequestContext = getRequestContext({ fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery });

        const responseData = filterResponseData(
          pendingRequestData,
          activeRequestData,
          responses.nestedTypeQuery as ResponseData,
          {
            active: activeRequestContext,
            pending: pendingRequestContext,
          }
        );

        expect(responseData).toMatchSnapshot();
      });
    });

    describe('when the directives match', () => {
      it('should return the correct data', () => {
        const pendingRequestData = getRequestData(parsedRequests.singleTypeQueryWithDirective);

        const pendingRequestContext = getRequestContext({
          fieldTypeMap: requestFieldTypeMaps.singleTypeQueryWithDirective,
        });

        const activeRequestData = getRequestData(parsedRequests.nestedTypeQueryWithDirectives);

        const activeRequestContext = getRequestContext({
          fieldTypeMap: requestFieldTypeMaps.nestedTypeQueryWithDirectives,
        });

        const responseData = filterResponseData(
          pendingRequestData,
          activeRequestData,
          responses.nestedTypeQuery as ResponseData,
          {
            active: activeRequestContext,
            pending: pendingRequestContext,
          }
        );

        expect(responseData).toMatchSnapshot();
      });
    });
  });

  describe('when there is a inline fragment in the pending query', () => {
    it('should return the correct data', () => {
      const pendingRequestData = getRequestData(parsedRequests.singleTypeQueryWithInlineFragment);

      const pendingRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.singleTypeQuery,
      });

      const activeRequestData = getRequestData(parsedRequests.nestedTypeQuery);

      const activeRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery,
      });

      const responseData = filterResponseData(
        pendingRequestData,
        activeRequestData,
        responses.nestedTypeQuery as ResponseData,
        {
          active: activeRequestContext,
          pending: pendingRequestContext,
        }
      );

      expect(responseData).toMatchSnapshot();
    });
  });

  describe('when there is a fragment spread in the new query', () => {
    it('should return the correct data', () => {
      const pendingRequestData = getRequestData(parsedRequests.singleTypeQueryWithFragmentSpread);

      const pendingRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.singleTypeQuery,
      });

      const activeRequestData = getRequestData(parsedRequests.nestedTypeQuery);

      const activeRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery,
      });

      const responseData = filterResponseData(
        pendingRequestData,
        activeRequestData,
        responses.nestedTypeQuery as ResponseData,
        {
          active: activeRequestContext,
          pending: pendingRequestContext,
        }
      );

      expect(responseData).toMatchSnapshot();
    });
  });

  describe('when there is are fragments in the active query', () => {
    it('should return the correct data', () => {
      const pendingRequestData = getRequestData(parsedRequests.nestedTypeQueryBasic);

      const pendingRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery,
      });

      const activeRequestData = getRequestData(parsedRequests.nestedTypeQueryWithFragments);

      const activeRequestContext = getRequestContext({
        fieldTypeMap: requestFieldTypeMaps.nestedTypeQuery,
      });

      const responseData = filterResponseData(
        pendingRequestData,
        activeRequestData,
        responses.nestedTypeQuery as ResponseData,
        {
          active: activeRequestContext,
          pending: pendingRequestContext,
        }
      );

      expect(responseData).toMatchSnapshot();
    });
  });
});
