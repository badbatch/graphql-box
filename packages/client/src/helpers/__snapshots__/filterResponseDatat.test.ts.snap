// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterResponseData when there are no fragments or directives or aliases in the pending query should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is a directive in the pending query when the directives do not match should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is a directive in the pending query when the directives match should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is a fragment spread in the new query should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is a inline fragment in the pending query should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is an alias in the pending query should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "fullName": "Facebook",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;

exports[`filterResponseData when there is are fragments in the active query should return the correct data 1`] = `
Object {
  "cacheMetadata": Map {},
  "data": Object {
    "organization": Object {
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "repositories": Object {
        "edges": Array [
          Object {
            "node": Object {
              "description": "Codemod is a tool/library to assist you with large-scale codebase refactors that can be partially automated but still require human oversight and occasional intervention. Codemod was developed at Facebook and released as open source.",
              "homepageUrl": "",
              "id": "MDEwOlJlcG9zaXRvcnkxNjU4ODM=",
              "name": "codemod",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
          Object {
            "node": Object {
              "description": "A virtual machine for executing programs written in Hack.",
              "homepageUrl": "https://hhvm.com",
              "id": "MDEwOlJlcG9zaXRvcnk0NTU2MDA=",
              "name": "hhvm",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
          Object {
            "node": Object {
              "description": "Python wrapper for RE2",
              "homepageUrl": "",
              "id": "MDEwOlJlcG9zaXRvcnk1NjU0MjY=",
              "name": "pyre2",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
          Object {
            "node": Object {
              "description": null,
              "homepageUrl": "http://ogp.me",
              "id": "MDEwOlJlcG9zaXRvcnk2MTkyNDA=",
              "name": "open-graph-protocol",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
          Object {
            "node": Object {
              "description": "Used to integrate Android apps with Facebook Platform.",
              "homepageUrl": "https://developers.facebook.com/docs/android",
              "id": "MDEwOlJlcG9zaXRvcnk2NTkzNDE=",
              "name": "facebook-android-sdk",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
          Object {
            "node": Object {
              "description": "Used to integrate the Facebook Platform with your iOS & tvOS apps.",
              "homepageUrl": "https://developers.facebook.com/docs/ios",
              "id": "MDEwOlJlcG9zaXRvcnk3Mzg0OTE=",
              "name": "facebook-objc-sdk",
              "owner": Object {
                "__typename": "Organization",
                "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "url": "https://github.com/facebook",
              },
            },
          },
        ],
      },
    },
  },
  "headers": Headers {
    Symbol(map): Object {
      "Cache-Control": Array [
        "public, max-age=5",
      ],
    },
  },
}
`;
